requests.go

```diff
@@ -0,0 +1,294 @@
+	}
+
+	// Add common headers
+	req = addHeadersToRequest(req, commonHeaders)
+
+	// Add custom headers
+	req = addHeadersToRequest(req, customHeaders)
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		b, err := io.ReadAll(resp.Body)
+		if err != nil {
+			log.Printf("failed to read response body: %s", err)
+		}
+		return fmt.Errorf("got HTTP status code %d with body: \n%s", resp.StatusCode, string(b))
+	}
+
+	if userData != nil {
+		err = json.NewDecoder(resp.Body).Decode(userData)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func loginStep(client *http.Client, portal *Portal, username, password string) error {
+	// GET login page at scheme+hostname+/portal/
+	hostname := portal.Hostname
+	idmHostname := portal.IDMHostname
+
+	headers := map[string][]string{
+		"Host":                      {hostname},
+		"upgrade-insecure-requests": {"1"},
+		"sec-fetch-site":            {"none"},
+		"sec-fetch-mode":            {"navigate"},
+		"sec-fetch-user":            {"?1"},
+		"sec-fetch-dest":            {"document"},
+	}
+
+	err := sendRequest(client, http.MethodGet, scheme+hostname+loginPagePath, headers, nil, nil)
+	if err != nil {
+		return err
+	}
+
+	// This request GETs the IDMSession Cookie.
+	headers = map[string][]string{
+		"Host":           {hostname},
+		"sec-fetch-site": {"same-origin"},
+		"sec-fetch-mode": {"cors"},
+		"sec-fetch-dest": {"empty"},
+		"pragma":         {"no-cache"},
+		"referer":        {scheme + hostname},
+	}
+
+	err = sendRequest(client, http.MethodGet, scheme+hostname+loginClearPath, headers, nil, nil)
+	if err != nil {
+		return err
+	}
+
+	// POST  scheme+hostname+/portal/login to get sessionToken used for oauth2 token
+	creds := loginDetails{
+		Username: username,
+		Password: password,
+		Options:  map[string]bool{"warnBeforePasswordExpired": true},
+	}
+
+	body, err := json.Marshal(creds)
+	if err != nil {
+		return err
+	}
+
+	headers = map[string][]string{
+		"Host":           {hostname},
+		"sec-fetch-site": {"same-origin"},
+		"sec-fetch-mode": {"cors"},
+		"sec-fetch-dest": {"empty"},
+		"pragma":         {"no-cache"},
+		"referer":        {scheme + hostname + "/portal/"},
+		"origin":         {hostname},
+	}
+
+	userData := &userData{}
+	err = sendRequest(client, http.MethodPost, scheme+hostname+loginSubmitPath, headers, body, userData)
+	if err != nil {
+		return err
+	}
+
+	if userData.SessionToken == "" {
+		return errors.New("missing sessionToken in response body; user might be locked out of portal")
+	}
+
+	// Start the oauth2 process
+	// GET scheme+idmHostname+/login/sessionCookieRedirect?token=&redirectUrl=scheme+hostname+/myportal/
+	// get the sessionToken from the response body of the POST /portal/login request and use as oauth2 token
+	token := userData.SessionToken
+	params := fmt.Sprintf("token=%s&redirectUrl=%s%s%s", token, scheme, hostname, oauth2RedirectUrlPath)
```

Comment:
You can use `https://pkg.go.dev/net/url#Values` to construct this for you and then you don't need to worry about escaping (which is missing here)

Reply:
Done.